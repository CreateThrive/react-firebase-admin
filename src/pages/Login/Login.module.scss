.errorMessage {
  margin: 0.7rem 0 0 0;
}
.socialButtons {
  flex-direction: column;
}
.socialButton {
  border-radius: 4px;
  padding: calc(0.375em - 1px) .75em calc(0.375em - 1px) .75em;
  margin: 5px;
  text-align: center;
}
.icon {
  margin-right: 5px;
}
