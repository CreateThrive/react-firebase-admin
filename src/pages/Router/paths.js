export default {
  ROOT: '/',
  LOGIN: '/login',
  USERS: '/users',
  ADD_USER: '/users/new',
  MODIFY_USER: '/users/:id',
  PROFILE: '/profile',
  RESET_PASSWORD: '/recover-password',
  SECTION: '/section',
  SUBMENU_1: '/submenu1',
  SUBMENU_2: '/submenu2',
  TEAMS: '/teams',
};
