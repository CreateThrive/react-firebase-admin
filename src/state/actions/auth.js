import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';

import { firebaseError, FIREBASE_RESPONSE } from 'utils';
import firebase from 'firebase.js';
import { clearUsersDataLogout } from './users';

export const AUTH_SIGN_IN_INIT = createAction('AUTH_SIGN_IN_INIT');
export const AUTH_SIGN_IN_FAIL = createAction('AUTH_SIGN_IN_FAIL');

export const AUTH_RESTORE_SESSION_INIT = createAction(
  'AUTH_RESTORE_SESSION_INIT'
);
export const AUTH_RESTORE_SESSION_SUCCESS = createAction(
  'AUTH_RESTORE_SESSION_SUCCESS'
);
export const AUTH_RESTORE_SESSION_FAIL = createAction(
  'AUTH_RESTORE_SESSION_FAIL'
);

export const AUTH_LOGOUT_INIT = createAction('AUTH_LOGOUT_INIT');
export const AUTH_LOGOUT_SUCCESS = createAction('AUTH_LOGOUT_SUCCESS');

export const AUTH_SET_PASSWORD_INIT = createAction('AUTH_SET_PASSWORD_INIT');
export const AUTH_SET_PASSWORD_SUCCESS = createAction(
  'AUTH_SET_PASSWORD_SUCCESS'
);
export const AUTH_SET_PASSWORD_FAIL = createAction('AUTH_SET_PASSWORD_FAIL');

export const AUTH_RESET_PASSWORD_INIT = createAction(
  'AUTH_RESET_PASSWORD_INIT'
);
export const AUTH_RESET_PASSWORD_SUCCESS = createAction(
  'AUTH_RESET_PASSWORD_SUCCESS'
);
export const AUTH_RESET_PASSWORD_FAIL = createAction(
  'AUTH_RESET_PASSWORD_FAIL'
);

export const AUTH_CLEAN_UP = createAction('AUTH_CLEAN_UP');

export const AUTH_FETCH_USER_DATA_INIT = createAction(
  'AUTH_FETCH_USER_DATA_INIT'
);
export const AUTH_FETCH_USER_DATA_SUCCESS = createAction(
  'AUTH_FETCH_USER_DATA_SUCCESS'
);
export const AUTH_FETCH_USER_DATA_FAIL = createAction(
  'AUTH_FETCH_USER_DATA_FAIL'
);

export const AUTH_CHANGE_PASSWORD_INIT = createAction(
  'AUTH_CHANGE_PASSWORD_INIT'
);
export const AUTH_CHANGE_PASSWORD_SUCCESS = createAction(
  'AUTH_CHANGE_PASSWORD_SUCCESS'
);
export const AUTH_CHANGE_PASSWORD_FAIL = createAction(
  'AUTH_CHANGE_PASSWORD_FAIL'
);

export const AUTH_UPDATE_USER_DATA = createAction('AUTH_UPDATE_USER_DATA');

export const AUTH_PROVIDER_INIT = createAction('AUTH_PROVIDER_INIT');

export const AUTH_PROVIDER_SUCCESS = createAction('AUTH_PROVIDER_SUCCESS');

export const AUTH_PROVIDER_FAIL = createAction('AUTH_PROVIDER_FAIL');

export const logout = () => {
  return async dispatch => {
    dispatch(AUTH_LOGOUT_INIT());

    dispatch(clearUsersDataLogout());
    await firebase.auth().signOut();

    dispatch(AUTH_LOGOUT_SUCCESS());
  };
};

export const verifyAuth = () => {
  return dispatch => {
    firebase.auth().onAuthStateChanged(user => {
      dispatch(AUTH_RESTORE_SESSION_INIT());

      if (user !== null) {
        return dispatch(AUTH_RESTORE_SESSION_SUCCESS());
      }

      dispatch(AUTH_RESTORE_SESSION_FAIL());
      return dispatch(logout());
    });
  };
};

export const fetchUserData = () => {
  return async dispatch => {
    dispatch(AUTH_FETCH_USER_DATA_INIT());

    const { uid } = firebase.auth().currentUser;

    let user;

    try {
      user = (
        await firebase
          .database()
          .ref(`users/${uid}`)
          .once('value')
      ).val();
    } catch (error) {
      dispatch(logout());
      return dispatch(AUTH_FETCH_USER_DATA_FAIL({ error }));
    }

    if (!user) {
      return dispatch(logout());
    }

    return dispatch(
      AUTH_FETCH_USER_DATA_SUCCESS({
        id: uid,
        ...user
      })
    );
  };
};

export const checkUserData = () => {
  return (dispatch, getState) => {
    const { id } = getState().auth.userData;

    if (!id) {
      dispatch(fetchUserData());
    }
  };
};

export const auth = (email, password) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_SIGN_IN_INIT());
    const { locale } = getState().preferences;
    try {
      await firebase.auth().signInWithEmailAndPassword(email, password);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SIGN_IN_FAIL({ error: errorMessage }));
    }

    const { emailVerified } = firebase.auth().currentUser;

    if (!emailVerified) {
      const errorMessage = firebaseError(
        FIREBASE_RESPONSE.USER_DISABLED,
        locale
      );
      return dispatch(AUTH_SIGN_IN_FAIL({ error: errorMessage }));
    }

    return dispatch(fetchUserData());
  };
};

export const setPassword = (email, password, url) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_SET_PASSWORD_INIT());
    const { locale } = getState().preferences;

    try {
      await firebase.auth().signInWithEmailLink(email, url);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SET_PASSWORD_FAIL({ error: errorMessage }));
    }

    const user = firebase.auth().currentUser;

    try {
      await user.updatePassword(password);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_SET_PASSWORD_FAIL({ error: errorMessage }));
    }

    dispatch(AUTH_SET_PASSWORD_SUCCESS());

    return dispatch(fetchUserData());
  };
};

export const resetPassword = email => {
  return async (dispatch, getState) => {
    dispatch(AUTH_RESET_PASSWORD_INIT());
    const { locale } = getState().preferences;

    try {
      await firebase.auth().sendPasswordResetEmail(email);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_RESET_PASSWORD_FAIL({ error: errorMessage }));
    }

    return dispatch(AUTH_RESET_PASSWORD_SUCCESS());
  };
};

export const authCleanUp = () => dispatch => dispatch(AUTH_CLEAN_UP());

export const changeUserPassword = (currentPassword, newPassword) => {
  return async (dispatch, getState) => {
    dispatch(AUTH_CHANGE_PASSWORD_INIT());
    const { locale } = getState().preferences;

    const user = firebase.auth().currentUser;

    const { email } = user;

    const credential = firebase.auth.EmailAuthProvider.credential(
      email,
      currentPassword
    );

    try {
      await user.reauthenticateWithCredential(credential);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(AUTH_CHANGE_PASSWORD_FAIL({ error: errorMessage }));
    }

    try {
      await user.updatePassword(newPassword);
    } catch (error) {
      const errorMessage = firebaseError(error, locale);
      toastr.error('', errorMessage);
      return dispatch(AUTH_CHANGE_PASSWORD_FAIL({ error: errorMessage }));
    }

    toastr.success('', 'Password changed successfully');
    return dispatch(AUTH_CHANGE_PASSWORD_SUCCESS());
  };
};

const authWithProvider = provider => {
  return async (dispatch, getState) => {
    dispatch(AUTH_PROVIDER_INIT());
    const { locale } = getState().preferences;
    let logInData;

    try {
      logInData = await firebase.auth().signInWithPopup(provider);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
    }
    const { user, additionalUserInfo } = logInData;

    const { uid } = firebase.auth().currentUser;

    const createdAt = new Date().toString();

    const { location } = additionalUserInfo.profile;

    const userData = {
      isAdmin: false,
      email: user.email,
      name: user.displayName,
      createdAt,
      logoUrl: user.photoURL,
      location: location?.name || null
    };

    let userFromDb;
    try {
      userFromDb = (
        await firebase
          .database()
          .ref(`users/${uid}`)
          .once('value')
      ).val();
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
    }

    if (!userFromDb) {
      try {
        await firebase
          .database()
          .ref(`users/${uid}`)
          .set(userData);
      } catch (error) {
        const errorMessage = firebaseError(error.code, locale);
        return dispatch(AUTH_PROVIDER_FAIL({ error: errorMessage }));
      }
    }

    return dispatch(
      AUTH_PROVIDER_SUCCESS({ id: uid, ...userData, ...userFromDb })
    );
  };
};

export const authFacebook = () => {
  const provider = new firebase.auth.FacebookAuthProvider();
  provider.addScope('email');
  return authWithProvider(provider);
};

export const authGoogle = () => {
  const provider = new firebase.auth.GoogleAuthProvider();
  provider.addScope('https://www.googleapis.com/auth/user.addresses.read');
  provider.addScope('https://www.googleapis.com/auth/userinfo.email');
  return authWithProvider(provider);
};

export const authMicrosoft = () => {
  const provider = new firebase.auth.OAuthProvider('microsoft.com');
  return authWithProvider(provider);
};
