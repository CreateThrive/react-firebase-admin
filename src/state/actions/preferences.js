import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
