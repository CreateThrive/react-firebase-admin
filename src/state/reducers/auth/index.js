import { createReducer } from 'redux-act';

import {
  AUTH_SIGN_IN_INIT,
  AUTH_SIGN_IN_FAIL,
  AUTH_FETCH_USER_DATA_INIT,
  AUTH_FETCH_USER_DATA_SUCCESS,
  AUTH_FETCH_USER_DATA_FAIL,
  AUTH_LOGOUT_INIT,
  AUTH_LOGOUT_SUCCESS,
  AUTH_RESTORE_SESSION_INIT,
  AUTH_RESTORE_SESSION_SUCCESS,
  AUTH_RESTORE_SESSION_FAIL,
  AUTH_SET_PASSWORD_INIT,
  AUTH_SET_PASSWORD_SUCCESS,
  AUTH_SET_PASSWORD_FAIL,
  AUTH_RESET_PASSWORD_INIT,
  AUTH_RESET_PASSWORD_SUCCESS,
  AUTH_RESET_PASSWORD_FAIL,
  AUTH_CLEAN_UP,
  AUTH_CHANGE_PASSWORD_INIT,
  AUTH_CHANGE_PASSWORD_SUCCESS,
  AUTH_CHANGE_PASSWORD_FAIL,
  AUTH_UPDATE_USER_DATA,
  AUTH_PROVIDER_FAIL,
  AUTH_PROVIDER_INIT,
  AUTH_PROVIDER_SUCCESS
} from 'state/actions/auth';

const initialState = {
  userData: {
    id: null,
    isAdmin: null
  },
  loading: false,
  error: null,
  restoring: false,
  restoringError: null,
  restoredPassword: false,
  changedPassword: false
};

export const authReducer = createReducer(
  {
    [AUTH_SIGN_IN_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_SIGN_IN_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_FETCH_USER_DATA_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_FETCH_USER_DATA_SUCCESS]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        isAdmin: payload.isAdmin,
        email: payload.email,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl,
        createdAt: payload.createdAt
      },
      loading: false,
      error: null
    }),
    [AUTH_FETCH_USER_DATA_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_LOGOUT_INIT]: () => ({ ...initialState }),
    [AUTH_LOGOUT_SUCCESS]: state => ({ ...state }),
    [AUTH_RESTORE_SESSION_INIT]: state => ({ ...state, restoring: true }),
    [AUTH_RESTORE_SESSION_SUCCESS]: state => ({
      ...state,
      restoring: false,
      restoringError: null
    }),
    [AUTH_RESTORE_SESSION_FAIL]: state => ({
      ...state,
      restoring: false,
      restoringError: true
    }),
    [AUTH_SET_PASSWORD_INIT]: state => ({ ...state, loading: true }),
    [AUTH_SET_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      error: null
    }),
    [AUTH_SET_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_RESET_PASSWORD_INIT]: () => ({
      ...initialState,
      loading: true
    }),
    [AUTH_RESET_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      error: null,
      restoredPassword: true
    }),
    [AUTH_RESET_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_CLEAN_UP]: state => ({
      ...state,
      error: null,
      changedPassword: false
    }),
    [AUTH_CHANGE_PASSWORD_INIT]: state => ({
      ...state,
      loading: true
    }),
    [AUTH_CHANGE_PASSWORD_SUCCESS]: state => ({
      ...state,
      loading: false,
      changedPassword: true
    }),
    [AUTH_CHANGE_PASSWORD_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    }),
    [AUTH_UPDATE_USER_DATA]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        email: state.userData.email,
        isAdmin: payload.isAdmin,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl || state.userData.logoUrl,
        createdAt: payload.createdAt
      }
    }),
    [AUTH_PROVIDER_INIT]: state => ({
      ...state,
      loading: true
    }),
    [AUTH_PROVIDER_SUCCESS]: (state, payload) => ({
      ...state,
      userData: {
        id: payload.id,
        isAdmin: payload.isAdmin,
        email: payload.email,
        name: payload.name,
        location: payload.location,
        logoUrl: payload.logoUrl,
        createdAt: payload.createdAt
      },
      error: null,
      loading: false
    }),
    [AUTH_PROVIDER_FAIL]: (state, payload) => ({
      ...state,
      loading: false,
      error: payload.error
    })
  },
  initialState
);
