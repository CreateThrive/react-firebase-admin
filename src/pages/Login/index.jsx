import React, { useEffect, useState, useCallback } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';
import classNames from 'classnames';
import 'bulma-social/bin/bulma-social.min.css';

import firebase from 'firebase.js';
import {
  auth,
  setPassword,
  authCleanUp,
  authFacebook,
  authGoogle,
  authMicrosoft
} from 'state/actions/auth';
import { useChangeHandler, useFormatMessage } from 'hooks';
import { inputValidations } from 'utils';
import paths from '../Router/paths';
import classes from './Login.module.scss';

const Login = () => {
  const { error, isAuth, loading, locale } = useSelector(
    state => ({
      error: state.auth.error,
      isAuth: !!state.auth.userData.id,
      loading: state.auth.loading,
      locale: state.preferences.locale
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const [authData, setAuthData] = useState({
    email: '',
    password: ''
  });

  const onChangeHandler = useChangeHandler(setAuthData);

  useEffect(() => {
    document.documentElement.classList.remove(
      'has-aside-left',
      'has-navbar-fixed-top'
    );
    return () => {
      document.documentElement.classList.add(
        'has-aside-left',
        'has-navbar-fixed-top'
      );
      dispatch(authCleanUp());
    };
  }, [dispatch]);

  const isEmailLink = firebase
    .auth()
    .isSignInWithEmailLink(window.location.href);

  const onSubmitHandler = event => {
    event.preventDefault();

    if (isEmailLink) {
      dispatch(
        setPassword(authData.email, authData.password, window.location.href)
      );
    } else {
      dispatch(auth(authData.email, authData.password));
    }
  };

  const iconsClassName = classNames('icon', classes.icon);

  const onFacebookHandler = useCallback(() => {
    dispatch(authFacebook());
  }, [dispatch, authFacebook]);

  const onGoogleHandler = useCallback(() => {
    dispatch(authGoogle());
  }, [dispatch, authGoogle]);

  const onMicrosoftHandler = useCallback(() => {
    dispatch(authMicrosoft());
  }, [dispatch, authMicrosoft]);

  const inputs = isEmailLink
    ? inputValidations(authData.email, authData.password, locale)
    : {
        email: {
          modifier: null,
          message: null
        },
        password: {
          modifier: null,
          message: null
        },
        canSubmit: false
      };

  const redirect = isAuth && <Redirect to={paths.ROOT} />;

  const setNewPasswordMessage = useFormatMessage('Login.setNewPassword');

  const loginMessage = useFormatMessage('Login.login');

  const setPasswordMessage = useFormatMessage('Login.setPassword');

  const forgotPasswordMessage = useFormatMessage('Login.forgotPassword');

  return (
    <section className="section hero is-fullheight is-error-section">
      {redirect}
      <div className="hero-body">
        <div className="container">
          <div className="columns is-centered">
            <div className="column is-two-fifths">
              <div className="card has-card-header-background">
                <header className="card-header">
                  <p className="card-header-title">
                    <span className="icon">
                      <i className="mdi mdi-lock default" />
                    </span>
                    <span>
                      {isEmailLink ? setNewPasswordMessage : loginMessage}
                    </span>
                  </p>
                </header>
                <div className="card-content">
                  <form onSubmit={onSubmitHandler}>
                    <div className="field">
                      <p className="label">{useFormatMessage('Login.email')}</p>
                      <div className="control is-clearfix">
                        <input
                          className={classNames('input', inputs.email.modifier)}
                          type="email"
                          name="email"
                          required
                          value={authData.email}
                          onChange={onChangeHandler}
                        />
                      </div>
                      {inputs.email.message && (
                        <p
                          className={classNames('help', inputs.email.modifier)}
                        >
                          {inputs.email.message}
                        </p>
                      )}
                    </div>
                    <div className="field">
                      <p className="label">
                        {useFormatMessage('Login.password')}
                      </p>
                      <div className="control is-clearfix">
                        <input
                          className={classNames(
                            'input',
                            inputs.password.modifier
                          )}
                          type="password"
                          name="password"
                          required
                          value={authData.password}
                          onChange={onChangeHandler}
                        />
                      </div>
                      {inputs.password.message && (
                        <p
                          className={classNames(
                            'help',
                            inputs.password.modifier
                          )}
                        >
                          {inputs.password.message}
                        </p>
                      )}
                    </div>
                    <br />
                    <div className="field is-grouped">
                      <div className="control">
                        <button
                          type="submit"
                          className={classNames('button', 'is-black', {
                            'is-loading': loading
                          })}
                          disabled={isEmailLink ? !inputs.canSubmit : false}
                        >
                          {isEmailLink ? setPasswordMessage : loginMessage}
                        </button>
                      </div>
                      {!isEmailLink && (
                        <div className="control">
                          <Link
                            to={paths.RESET_PASSWORD}
                            className="button is-outlined"
                          >
                            {forgotPasswordMessage}
                          </Link>
                        </div>
                      )}
                    </div>
                    {error && (
                      <p
                        className={classNames(
                          'has-text-danger',
                          classes.errorMessage
                        )}
                      >
                        {error}
                      </p>
                    )}
                  </form>
                  <hr />
                  <div
                    className={classNames(
                      'field',
                      'is-grouped',
                      classes.socialButtons
                    )}
                  >
                    <a
                      className={classNames(
                        'is-facebook',
                        classes.socialButton
                      )}
                      onClick={onFacebookHandler}
                    >
                      <span className={iconsClassName}>
                        <i className="mdi mdi-facebook" />
                      </span>
                      <span>{useFormatMessage('Login.facebook')}</span>
                    </a>
                    <a
                      className={classNames('is-google', classes.socialButton)}
                      onClick={onGoogleHandler}
                    >
                      <span className={iconsClassName}>
                        <i className="mdi mdi-google" />
                      </span>
                      <span>{useFormatMessage('Login.google')}</span>
                    </a>
                    <a
                      className={classNames(
                        'is-microsoft',
                        classes.socialButton
                      )}
                      onClick={onMicrosoftHandler}
                    >
                      <span className={iconsClassName}>
                        <i className="mdi mdi-microsoft" />
                      </span>
                      <span>{useFormatMessage('Login.microsoft')}</span>
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  );
};

export default Login;
