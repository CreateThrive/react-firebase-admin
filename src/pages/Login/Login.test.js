import React from 'react';
import { Redirect } from 'react-router-dom';
import * as reactRedux from 'react-redux';

import * as authActions from 'state/actions/auth';
import paths from '../Router/paths';
import Login from '.';

describe('<Login /> rendering', () => {
  it('should render without crashing', () => {
    const { component } = mountWithProviders(<Login />)({
      auth: {
        userData: {}
      }
    });

    expect(component).toMatchSnapshot();
  });

  it('should display an error message when there is an error', () => {
    const { component } = mountWithProviders(<Login />)({
      auth: {
        userData: {},
        error: 'sample error'
      }
    });

    expect(component.find('.has-text-danger').length).toBe(1);
  });

  it('should redirect to /home when the user is authenticated', () => {
    const { component } = mountWithProviders(<Login />)({
      auth: {
        userData: {
          id: 'some userId'
        }
      }
    });

    expect(component.contains(<Redirect to={paths.ROOT} />)).toEqual(true);
  });
});

const dispatchMock = jest.fn();

beforeEach(() => {
  jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock);
  jest.spyOn(authActions, 'authFacebook').mockImplementation(jest.fn);
  jest.spyOn(authActions, 'authGoogle').mockImplementation(jest.fn);
  jest.spyOn(authActions, 'authMicrosoft').mockImplementation(jest.fn);
});

it('should dispatch authFacebook action when the user tries to log in with Facebook', () => {
  const { component } = mountWithProviders(<Login />)({
    auth: {
      userData: {}
    }
  });

  component.find('.is-facebook').simulate('click');

  expect(authActions.authFacebook).toHaveBeenCalled();
});

it('should dispatch authGoogle action when the user tries to log in with Google', () => {
  const { component } = mountWithProviders(<Login />)({
    auth: {
      userData: {}
    }
  });

  component.find('.is-google').simulate('click');

  expect(authActions.authGoogle).toHaveBeenCalled();
});

it('should dispatch authMicrosoft action when the user tries to log in with Microsoft', () => {
  const { component } = mountWithProviders(<Login />)({
    auth: {
      userData: {}
    }
  });

  component.find('.is-microsoft').simulate('click');

  expect(authActions.authMicrosoft).toHaveBeenCalled();
});
