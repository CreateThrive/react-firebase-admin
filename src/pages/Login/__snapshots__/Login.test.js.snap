// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> rendering should render without crashing 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en"
  messages={
    Object {
      "Aside.dropdownMenu": "Dropdown Menu",
      "Aside.home": "Home",
      "Aside.submenu1": "Submenu 1",
      "Aside.submenu2": "Submenu 2",
      "Aside.users": "Users",
      "ChangePassword.changePassword": "Change Password",
      "ChangePassword.confirmPassword": "Confirm Password",
      "ChangePassword.currentPassword": "Current Password",
      "ChangePassword.insecurePassword": "Insecure password",
      "ChangePassword.matchPassword": "Passwords match",
      "ChangePassword.newPassword": "New Password",
      "ChangePassword.notMatchPassword": "Passwords do not match",
      "ChangePassword.safePassword": "Safe password",
      "ChangePassword.samePassword": "The new password and the current one cannot be the same",
      "ChangePassword.submits": "Submit",
      "Home.content": "Home content",
      "Home.home": "Home",
      "Login.email": "E-mail Address",
      "Login.facebook": "Continue with Facebook",
      "Login.forgotPassword": "Forgot Password?",
      "Login.google": "Continue with Google",
      "Login.login": "Login",
      "Login.microsoft": "Continue with Microsoft",
      "Login.password": "Password",
      "Login.setNewPassword": "Set your new password",
      "Login.setPassword": "Set Password",
      "NavBar.Language": "English",
      "NavBar.logOut": "Log Out",
      "NavBar.profile": "Profile",
      "NotFound.404": "Error 404: page not found",
      "NotFound.back": "Go Back",
      "NotFound.url": "The requested URL {url} was not found",
      "Profile.profile": "Profile",
      "ResetPassword.back": "Back",
      "ResetPassword.email": "E-mail Address",
      "ResetPassword.emailRegistration": "E-mail used for registration",
      "ResetPassword.recoverEmail": "We have sent you an email to {mail} so you can recover your account.",
      "ResetPassword.recovery": "Password Recovery",
      "ResetPassword.resetLink": "Send Reset Link",
      "Section.content": "Section content",
      "Section.section": "Section",
      "Submenu.content": "Submenu content",
      "Submenu.submenu": "Submenu",
      "Table.perPage": " per page",
      "User.editUser": "Edit User",
      "User.newUser": "New User",
      "UserForm.admin": "Admin",
      "UserForm.created": "Created",
      "UserForm.email": "E-mail",
      "UserForm.goBack": "Go Back",
      "UserForm.invalidEmail": "Invalid E-mail",
      "UserForm.location": "Location",
      "UserForm.logo": "Logo",
      "UserForm.name": "Name",
      "UserForm.pickAnotherFile": "Pick another file",
      "UserForm.pickFile": "Pick a file",
      "UserForm.submit": "Submit",
      "UserForm.userInfo": "User Information",
      "UserForm.userPreview": "User Preview",
      "Users.admin": "Admin",
      "Users.cancel": "Cancel",
      "Users.confirm": "Confirm action",
      "Users.created": "Created",
      "Users.delete": "Delete",
      "Users.email": "Email",
      "Users.location": "Location",
      "Users.name": "Name",
      "Users.newUser": "New User",
      "Users.permDelete": "This will permanently delete the user. Action can not be undone.",
      "Users.search": "Search:",
      "Users.users": "Users",
      "auth/account-exists-with-different-credential": "Email on another sign in method",
      "auth/cancelled-popup-request": "Too many pop ups, just one pop up is allowed",
      "auth/email-already-exists": "Email already in use",
      "auth/expired-action-code": "The invitation link has expired, get in touch with your administrator",
      "auth/invalid-action-code": "The invitation link has expired, get in touch with your administrator",
      "auth/invalid-email": "Email is invalid",
      "auth/popup-blocked": "Pop up blocked by the browser, please enable pop ups for this page",
      "auth/popup-closed-by-user": "Operation cancelled, pop up was closed",
      "auth/too-many-requests": "Too many attempts made, try again later",
      "auth/user-disabled": "User disabled",
      "auth/wrong-password": "Invalid credentials",
      "storage/quota-exceeded": "Internal server error, get in touch with your administrator",
      "storage/unauthenticated": "Unauthenticated, please authenticate and try again",
      "storage/unauthorized": "Unauthorized, you are not authorized to perform this action",
      "utils.default": "Unknown error, get in touch with your administrator",
      "utils.invalidEmail": "Invalid email",
      "utils.safePassword": "Safe password",
      "utils.unsafePassword": "Unsafe password",
    }
  }
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <Login>
          <section
            className="section hero is-fullheight is-error-section"
          >
            <div
              className="hero-body"
            >
              <div
                className="container"
              >
                <div
                  className="columns is-centered"
                >
                  <div
                    className="column is-two-fifths"
                  >
                    <div
                      className="card has-card-header-background"
                    >
                      <header
                        className="card-header"
                      >
                        <p
                          className="card-header-title"
                        >
                          <span
                            className="icon"
                          >
                            <i
                              className="mdi mdi-lock default"
                            />
                          </span>
                          <span>
                            Login
                          </span>
                        </p>
                      </header>
                      <div
                        className="card-content"
                      >
                        <form
                          onSubmit={[Function]}
                        >
                          <div
                            className="field"
                          >
                            <p
                              className="label"
                            >
                              E-mail Address
                            </p>
                            <div
                              className="control is-clearfix"
                            >
                              <input
                                className="input"
                                name="email"
                                onChange={[Function]}
                                required={true}
                                type="email"
                                value=""
                              />
                            </div>
                          </div>
                          <div
                            className="field"
                          >
                            <p
                              className="label"
                            >
                              Password
                            </p>
                            <div
                              className="control is-clearfix"
                            >
                              <input
                                className="input"
                                name="password"
                                onChange={[Function]}
                                required={true}
                                type="password"
                                value=""
                              />
                            </div>
                          </div>
                          <br />
                          <div
                            className="field is-grouped"
                          >
                            <div
                              className="control"
                            >
                              <button
                                className="button is-black"
                                disabled={false}
                                type="submit"
                              >
                                Login
                              </button>
                            </div>
                            <div
                              className="control"
                            >
                              <Link
                                className="button is-outlined"
                                to="/recover-password"
                              >
                                <LinkAnchor
                                  className="button is-outlined"
                                  href="/recover-password"
                                  navigate={[Function]}
                                >
                                  <a
                                    className="button is-outlined"
                                    href="/recover-password"
                                    onClick={[Function]}
                                  >
                                    Forgot Password?
                                  </a>
                                </LinkAnchor>
                              </Link>
                            </div>
                          </div>
                        </form>
                        <hr />
                        <div
                          className="field is-grouped socialButtons"
                        >
                          <a
                            className="is-facebook socialButton"
                            onClick={[Function]}
                          >
                            <span
                              className="icon icon"
                            >
                              <i
                                className="mdi mdi-facebook"
                              />
                            </span>
                            <span>
                              Continue with Facebook
                            </span>
                          </a>
                          <a
                            className="is-google socialButton"
                            onClick={[Function]}
                          >
                            <span
                              className="icon icon"
                            >
                              <i
                                className="mdi mdi-google"
                              />
                            </span>
                            <span>
                              Continue with Google
                            </span>
                          </a>
                          <a
                            className="is-microsoft socialButton"
                            onClick={[Function]}
                          >
                            <span
                              className="icon icon"
                            >
                              <i
                                className="mdi mdi-microsoft"
                              />
                            </span>
                            <span>
                              Continue with Microsoft
                            </span>
                          </a>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </section>
        </Login>
      </Provider>
    </Router>
  </BrowserRouter>
</IntlProvider>
`;
